x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/use_prebuilt_rules_upgrade_state.ts

```diff
@@ -0,0 +1,116 @@
+
+export interface RuleUpgradeState extends RuleUpgradeInfoForReview {
+  /**
+   * Rule containing desired values users expect to see in the upgraded rule.
+   */
+  finalRule: DiffableRule;
+  /**
+   * Indicates whether there are conflicts blocking rule upgrading.
+   */
+  hasUnresolvedConflicts: boolean;
+}
+export type RulesUpgradeState = Record<RuleSignatureId, RuleUpgradeState>;
+export type SetFieldResolvedValueFn<
+  FieldName extends keyof DiffableAllFields = keyof DiffableAllFields
+> = (params: {
+  ruleId: RuleObjectId;
+  fieldName: FieldName;
+  resolvedValue: DiffableAllFields[FieldName];
+}) => void;
+
+type RuleResolvedConflicts = Partial<DiffableAllFields>;
+type RulesResolvedConflicts = Record<string, RuleResolvedConflicts>;
+
+interface UseRulesUpgradeStateResult {
+  rulesUpgradeState: RulesUpgradeState;
+  setFieldResolvedValue: SetFieldResolvedValueFn;
+}
+
+export function usePrebuiltRulesUpgradeState(
+  ruleUpgradeInfos: RuleUpgradeInfoForReview[]
+): UseRulesUpgradeStateResult {
+  const [rulesResolvedConflicts, setRulesResolvedConflicts] = useState<RulesResolvedConflicts>({});
+  const setFieldResolvedValue = useCallback((...[params]: Parameters<SetFieldResolvedValueFn>) => {
+    setRulesResolvedConflicts((prevRulesResolvedConflicts) => ({
+      ...prevRulesResolvedConflicts,
+      [params.ruleId]: {
+        ...(prevRulesResolvedConflicts[params.ruleId] ?? {}),
+        [params.fieldName]: params.resolvedValue,
+      },
+    }));
+  }, []);
+  const rulesUpgradeState = useMemo(() => {
+    const state: RulesUpgradeState = {};
+
+    for (const ruleUpgradeInfo of ruleUpgradeInfos) {
+      state[ruleUpgradeInfo.rule_id] = {
+        ...ruleUpgradeInfo,
+        finalRule: calcFinalDiffableRule(
+          ruleUpgradeInfo,
+          rulesResolvedConflicts[ruleUpgradeInfo.id] ?? {}
+        ),
+        hasUnresolvedConflicts:
+          calcUnresolvedConflicts(
+            ruleUpgradeInfo,
+            rulesResolvedConflicts[ruleUpgradeInfo.id] ?? {}
+          ) > 0,
+      };
+    }
+
+    return state;
+  }, [ruleUpgradeInfos, rulesResolvedConflicts]);
+
+  return {
+    rulesUpgradeState,
+    setFieldResolvedValue,
+  };
+}
+
+function calcFinalDiffableRule(
+  ruleUpgradeInfo: RuleUpgradeInfoForReview,
+  ruleResolvedConflicts: RuleResolvedConflicts
+): DiffableRule {
+  return {
+    ...convertRuleToDiffable(ruleUpgradeInfo.target_rule),
+    ...ruleResolvedConflicts,
+  } as DiffableRule;
+}
+
+function calcUnresolvedConflicts(
+  ruleUpgradeInfo: RuleUpgradeInfoForReview,
+  ruleResolvedConflicts: RuleResolvedConflicts
+): number {
+  const fieldNames = Object.keys(ruleUpgradeInfo.diff.fields) as Array<
+    keyof typeof ruleUpgradeInfo.diff.fields
+  >;
+  const fieldNamesWithNonSolvableConflicts = fieldNames.filter(
+    (fieldName) =>
+      ruleUpgradeInfo.diff.fields[fieldName]?.conflict === ThreeWayDiffConflict.NON_SOLVABLE
+  );
```

Comment:
We spoke about this today, but posting here for transparency: this should include the fields with `SOLVABLE` conflicts, which should be accepted by the user as well in order to unblock the upgrade of the rule.

Reply:
Yes. I changed the implementation to match our discussion results.